src/app/mesh/qgsmaptooleditmeshframe.cpp

```diff
@@ -2717,7 +2790,33 @@ void QgsMapToolEditMeshFrame::addVertex(
     zValue = QgsMeshLayerUtils::interpolateFromVerticesData( v1, v2, v3, v1.z(), v2.z(), v3.z(), mapPoint );
   }
   else
-    zValue = currentZValue();
+  {
+    if ( mZValueWidget->getZFromProjectElevationEnabled() )
+    {
+      const QgsAbstractTerrainProvider *terrainProvider = QgsProject::instance()->elevationProperties()->terrainProvider();
+      const QgsCoordinateTransform transformation = QgsCoordinateTransform( mCurrentLayer->crs(), terrainProvider->crs(), QgsProject::instance() );
+
+      try
+      {
+        const QgsPointXY point = transformation.transform( effectivePoint.x(), effectivePoint.y() );
+        zValue = terrainProvider->heightAt( point.x(), point.y() );
+      }
+      catch ( const QgsCsException & )
+      {
+        zValue = std::numeric_limits<double>::quiet_NaN();
+      }
```

Comment:
@uclaros unlike what we discussed last week, this cannot be moved up in the hierarchy of those `if else` statements.

I believe that this settings should only affect newly added vertices that are isolated. We probably do not want to affect behavior when adding vertices on edges and in faces (those have Z interpolated from the MESH vertices itself) and that is what the previous `if else` statements do. We only what to affect the `else` branch which are the isolated vertices. So I believe that this should be correct.